- Description: zfcp: Infinite retry of SCSI commands.
  Symptom:     System hangs.
  Problem:     If a machine check indicates that an FCP adapter is not available
               anymore the zfcp lldd did not shutdown the adapter internally.
               Under certain circumstances this can lead to a state where SCSI
               commands are always returned with return code
               SCSI_MLQUEUE_HOST_BUSY to the SCSI midlayer. The result is that
               the commands will be retried infinitely.
  Solution:    Shutdown adapter internally.
  Problem-ID:  -
  Archs:       s390-31, s390-64

--- linux-2.5/drivers/s390/scsi/zfcp_ccw.c	26 Mar 2004 07:00:32 -0000	1.52
+++ linux-2.5/drivers/s390/scsi/zfcp_ccw.c	29 Apr 2004 08:33:58 -0000	1.52.2.1
@@ -26,7 +26,7 @@
  * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
  */
 
-#define ZFCP_CCW_C_REVISION "$Revision: 1.52 $"
+#define ZFCP_CCW_C_REVISION "$Revision: 1.52.2.1 $"
 
 #include <linux/init.h>
 #include <linux/module.h>
@@ -232,14 +232,19 @@
 	case CIO_GONE:
 		ZFCP_LOG_NORMAL("adapter %s: device gone\n",
 				zfcp_get_busid_by_adapter(adapter));
+		debug_text_event(adapter->erp_dbf,1,"dev_gone");
+		zfcp_erp_adapter_shutdown(adapter, 0);
 		break;
 	case CIO_NO_PATH:
 		ZFCP_LOG_NORMAL("adapter %s: no path\n",
 				zfcp_get_busid_by_adapter(adapter));
+		debug_text_event(adapter->erp_dbf,1,"no_path");
+		zfcp_erp_adapter_shutdown(adapter, 0);
 		break;
 	case CIO_OPER:
 		ZFCP_LOG_NORMAL("adapter %s: operational again\n",
 				zfcp_get_busid_by_adapter(adapter));
+		debug_text_event(adapter->erp_dbf,1,"dev_oper");
 		zfcp_erp_modify_adapter_status(adapter,
 					       ZFCP_STATUS_COMMON_RUNNING,
 					       ZFCP_SET);
@@ -247,6 +252,7 @@
 					ZFCP_STATUS_COMMON_ERP_FAILED);
 		break;
 	}
+	zfcp_erp_wait(adapter);
 	up(&zfcp_data.config_sema);
 	return 1;
 }

